src/main/java/org/mskcc/smile/service/impl/ValidRequestCheckerImpl.java

```diff
@@ -126,13 +98,44 @@ public String getFilteredValidRequestJson(String requestJson) throws IOException
         return null;
     }
 
+    @Override
+    public Boolean isValidPromotedRequest(String requestJson) throws JsonMappingException,
+            JsonProcessingException, IOException {
+        // first check if request-level metadata is valid
+        if (!hasValidRequestLevelMetadata(requestJson)) {
+            return Boolean.FALSE;
+        }
+
+        // verify that request has 'samples' json field
+        if (!requestHasSamples(requestJson)) {
+            return Boolean.FALSE;
+        }
+
+        Map<String, Object> requestJsonMap = mapper.readValue(requestJson, Map.class);
+        Object[] sampleList = mapper.convertValue(requestJsonMap.get("samples"),
+                Object[].class);
+        for (Object sample: sampleList) {
+            Map<String, String> sampleMap = mapper.convertValue(sample, Map.class);
+            // sample should have at least an igoId/primaryId and at least a cmoPatientId/normalizedPatientId
+            if (!hasIgoId(sampleMap) || !(hasCmoPatientId(sampleMap) || hasNormalizedPatientId(sampleMap))) {
+                LOG.warn("Sample is missing one or a combination of the following: igoId or primaryId, "
+                        + "cmoPatientId or cmoSampleIdFields --> normalizedPatientId - this information "
+                        + "must be added for promoted requests & samples");
+                requestStatusLogger.logRequestStatus(requestJson,
+                    RequestStatusLogger.StatusType.PROMOTED_SAMPLES_MISSING_IDS);
+                return Boolean.FALSE;
+            }
+        }
+        return Boolean.TRUE;
+    }
+
     /**
      * Evaluates request metadata and returns a boolean based on whether the request data
      * passes all sanity checks.
      * @throws IOException
      */
     @Override
-    public Boolean isValidRequestMetadataJson(String requestJson)
+    public Boolean hasValidRequestLevelMetadata(String requestJson)
```

Comment:
I understand this is called from isValidPromotedRequest, but is this also called for a newRequest?  If so, do we have to check for fastQ paths?  Is that checked somewhere?

Reply:
we can add that